src/test-util.ts

```diff
@@ -72,91 +69,103 @@ type Response = {
 
 type LogResponse = {
   level: string;
-  message: string;
+  data: unknown[];
 };
 
 export async function initializeNewTab(event: MessageEvent): Promise<void> {
   const {tabId, path} = event.data;
+  const {opener} = window;
+  if (opener === null) {
+    return;
+  }
 
   ['error', 'warn', 'info', 'debug', 'log'].forEach(level => {
     // eslint-disable-next-line @typescript-eslint/ban-ts-comment
     // @ts-ignore No index signature of type 'string' was found on 'Console'.
     console[level] = (...data: unknown[]) => {
-      window.opener.postMessage(
-        {tabId, id: 'log', result: {level, message: data.join(' ')}},
-        window.origin,
-      );
+      opener.postMessage({tabId, id: 'log', result: {level, data}}, origin);
     };
   });
 
-  // eslint-disable-next-line @typescript-eslint/no-empty-function
-  const AsyncFunction = Object.getPrototypeOf(async function () {}).constructor;
   let module: typeof import('./test-util');
   if (path !== undefined) {
     try {
-      module = await import(window.origin + '/' + path);
+      module = await import(new URL('/' + path, location.href).toString());
     } catch (e) {
       console.log('Error importing ' + path + ':', e);
     }
   }
 
   const handle = async (event: MessageEvent) => {
-    if (event.source === null || event.source != window.opener) {
+    const {source, data} = event;
+    const {id, expr} = data;
+    const response: Response = {tabId, id};
+
+    if (source === null || source !== opener) {
       return;
     }
-    const {id, expr} = event.data;
-    const response: Response = {tabId, id};
     if (expr) {
       try {
-        response.result = await AsyncFunction(
+        response.result = await Function(
           'module',
-          '"use strict"; Object.assign(self, module); {' + expr + '}',
+          `"use strict";
+           Object.assign(self, module);
+           return (async () => { ${expr} })();`,
         )(module);
       } catch (e) {
         response.error = e;
       }
     }
-    event.source.postMessage(response);
+    opener.postMessage(response, origin);
   };
 
-  window.addEventListener('message', handle, false);
+  addEventListener('message', handle, false);
   void handle(event);
 }
 
+function withTimeout<T = void>(promise: Promise<T>, ms: number) {
+  const timeout = new Promise((_resolve, reject) => {
+    setTimeout(() => reject(new Error(`Timeout after ${ms}ms.`)), ms);
```

Comment:
you could clearTimeout when the other one is done but this is just test code so no big deal